### DISCUSSION TITLE
Deleting Data from Javers-Tables

### DISCUSSION THREAD

**Discussion by Nightbl00d** [javers/javers#1357] [Q&A]
Hello Community,
we have the problem that the Javers Data is getting big over the years. So we decided to create a job that deletes old commits and Snapshots from our Database. I am doing some research and testing right now and I found something interesting.

I have an entity, let's say, "address" and an entity "Person" with "address" as Attribute. There are also "sub-entities" for the address. The sub-entities are connected to the address through the owner_id_fk from the table jv_global_id.

First I simply deleted all the Snapshots and Commits. So the "global_ids" are still left. There is no connection between Address- and Person-"global_ids" over the owner_fk-column in jv_global_id. So I guess there is no connection at all. 
Now If I change the Address of one Person in our Springboot-Application a new second global_id for the address (with the same local_id as the existing global_id) is created. The new Snapshot references to the new global_id. But there are new sub-entities-entries aswell pointing at the old gloabl_id. So there are doubled "sub-entries" pointing at the old global_id. There is no Snapshot for Person. From that point the javers-history isn't working properly.

Then I tried to delete global_ids, that aren't used in any snapshots after I deleted the old Snapshots. Now if I changed the address of a person a new global_id with all the sub-entries are created and also a correct Snapshot of the "address". But there is no Snapshot for the Person. So there is no easy way to show the user the change of the persons address. If I add the Persons global_id manually with the fitting local_id the history will work again.

If I keep the initial-Snapshots and the global_id in the database everything still works fine. But we haven't decided yet if we want to keep the initial-Snapshots. I hope it is understandable what I am trying to say. Maybe I can add a diagram of what is happening.

Can someone explain why there are doubled global_ids in the first solution. Maybe javers always create new global_ids for the first Snapshot of an entity? 
And Is there a way to configure Javers to create the "Person"-Snapshot when creating an initial "Address"?
Has someone a different Idea? Maybe I am missing something.

Best Regards,
Michael

**Comment by bartoszwalacik** [comment]
Hi @Nightbl00d , first of all, Javers always reuses rows in `jv_global_id`. If you end up with more than one row in `jv_global_id` targeting the same GlobalId value -- it means data corruption. If you can provide a full reproduction scenario which leads to this situation, please report it.
I believe that the right way of deleting data from Javers tables is described here https://github.com/javers/javers/discussions/1352

**Reply by Nightbl00d** [reply]
Hi @bartoszwalacik , thank you for the reply. 

I deleted all Snapshots and kept the global_id. After I changed something on the object without initial-Snapshot a second global_id with the samel local_id appeared after changing something on the object. 

On my second try, I deleted the global_id and the Snapshots. Now the method "javers.findChanges()" gives back an error after adding changes. I can imagine it is because of the missing Initial-Snapshot.  

Have you ever tested if JaVers is working properly after deleting rows with your procedure? 

EDIT: I found out that the problem with "javers.findChanges()" occurs if there is no INITIAL-Entry in the Snapshot-Table, but an UPDATE-Entry. Perhaps JaVers is trying to compare the UPDATE with the INITIAL-State?? The Error says: "argument should not be null"

**Comment by bartoszwalacik** [comment]
Hi @Nightbl00d , deleting Data from Javers tables is not officially supported, happy you are investigating this. If you can suggest a better procedure than described here https://github.com/javers/javers/discussions/1352, please do

**Reply by 7erg** [reply]
Greetings!

Thank you for the suggestions in #1352.

I’ve tried to delete old records, but now I’m encountering an exception when searching for history via the findChanges method for an entity that no longer has the oldest snapshots:

```
java.lang.IllegalArgumentException: argument should not be null
    at org.javers.common.validation.Validate.argumentsAreNotNull(Validate.java:33)
    at org.javers.core.graph.ObjectNode.<init>(ObjectNode.java:30)
    at org.javers.core.snapshot.SnapshotNode.<init>(SnapshotNode.java:15)
    at org.javers.core.snapshot.SnapshotDiffer.snapshotGraph(SnapshotDiffer.java:73)
    at org.javers.core.snapshot.SnapshotDiffer.addChanges(SnapshotDiffer.java:68)
    at org.javers.core.snapshot.SnapshotDiffer.calculateDiffs(SnapshotDiffer.java:48)
    at org.javers.repository.api.JaversExtendedRepository.getChangesIntroducedBySnapshots(JaversExtendedRepository.java:181)
    at org.javers.repository.api.JaversExtendedRepository.getChangeHistory(JaversExtendedRepository.java:40)
    at org.javers.repository.jql.ChangesQueryRunner.queryForChanges(ChangesQueryRunner.java:30)
    at org.javers.repository.jql.QueryRunner.queryForChanges(QueryRunner.java:46)
    at org.javers.core.JaversCore.findChanges(JaversCore.java:206)
    at org.javers.spring.transactions.JaversTransactionalDecorator.findChanges(JaversTransactionalDecorator.java:143)
...
```
Would a change in `SnapshotDiffer` be the right approach to support missing snapshots?

**Reply by bartoszwalacik** [reply]
I guess so
